cmd/noms/diff/print_diff.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package diff
+
+import (
+	"io"
+
+	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/noms/go/util/writers"
+	"github.com/dustin/go-humanize"
+)
+
+type prefixOp string
+
+const (
+	ADD = "+   "
+	DEL = "-   "
+)
+
+type (
+	printFunc func(w io.Writer, op prefixOp, key, val types.Value) error
+)
+
+// PrintDiff writes a textual reprensentation of the diff from |v1| to |v2|
+// to |w|. If |leftRight| is true then the left-right diff is used for ordered
+// sequences - see Diff vs DiffLeftRight in Set and Map.
+func PrintDiff(w io.Writer, v1, v2 types.Value, leftRight bool) (err error) {
+	// In the case where the diff involves two simple values, just print out the
+	// diff and return. This is needed because the code below assumes that the
+	// values being compared have a parent.
+	if !shouldDescend(v1, v2) {
+		line(w, DEL, nil, v1)
+		return line(w, ADD, nil, v2)
+	}
+
+	dChan := make(chan Difference, 128)
+	stopChan := make(chan struct{})
+	stopDiff := func() {
+		close(stopChan)
+		for range dChan {
+		}
+	}
+
+	// From here on, we can assume that every Difference will have at least one
+	// element in the Path
+	differ := Differ{DiffChan: dChan, StopChan: stopChan, LeftRight: leftRight}
+	go func() {
+		differ.Diff(v1, v2)
+		close(dChan)
+	}()
+
+	var lastParentPath types.Path
+	wroteHdr := false
+	firstTime := true
+
+	for d := range dChan {
+		parentPath := d.Path[:len(d.Path)-1]
+		parentPathChanged := !parentPath.Equals(lastParentPath)
+		lastParentPath = parentPath
+		if parentPathChanged && wroteHdr {
+			writeFooter(w, &wroteHdr)
```

Comment:
You need to make sure that `err` is assigned if writing fails. If you have a chain of write calls, just assigning on the last one should be enough:
```
write()
write()
err = write()
```
but here all of the writes are behind conditionals, so I think you need to assign on every call. I.e. `err = writeFooter`. Same everywhere else basically (and writeHeader etc).